[JSON]Add sort video items by original title - SortByOriginalTitle

JimmyS83 opened the pull request:
This is Matrix version of a (now closed) PR originally submitted for Leia - https://github.com/xbmc/xbmc/pull/19434

## Description
Adds sort method, according `originaltitle` variable/infolabel, with fallback to standard `sorttitle`, `title`.

## Motivation and Context
Most of (newer) scrapers (such as default tmdb) supports scraping item `Title` in user’s choice language. They are able to store also original title into `originaltitle` variable (infolabel).

It could be displayed by skin in lists, etc. User can choose, which title he wants in lists, dynamically, even both of them for example.

The problem is, that Kodi lacks function to sort list according originaltitle. It supports only `Title`, and `Sorttitle` parameters. `Sorttitle` is written to DB during scraping, so its not dynamic parameter at all.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):
Example on dummy data, where both titles are visible to make a point more easilly: https://streamable.com/kgvjj9

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

JimmyS83 committed:
Add sorting of video items by original title

fuzzard commented:
You will actually need to PR this against master branch.
Our process is anything new must be merged into master, and then a backport may be considered for the previous version after this.
Having said that, we generally don't backport features, only fixes, but I'll let @DaveTBlake be the decider on this one

wsnipex commented:
jenkins build this please

JimmyS83 commented:
any update?

Montellese commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -109,7 +109,7 @@
                   "playlist", "episode", "season", "totalepisodes", "watchedepisodes", "tvshowstatus", "tvshowtitle",
                   "sorttitle", "productioncode", "mpaa", "studio", "dateadded", "lastplayed", "playcount", "listeners",
-                  "bitrate", "random", "totaldiscs", "originaldate", "bpm" ]
+                  "bitrate", "random", "totaldiscs", "originaldate", "bpm", "originaltitle" ]
```
You also need to bump the minor part of the JSON API version.

And please fix the commit message. It should at least mention `JSON-RPC` and `sort by original title`.

JimmyS83 replied:
Better now?

Montellese replied:
Yes as long as you squash the three commits concerning JSON-RPC into one :+1:

JimmyS83 replied:
Done :)

Montellese reviewed (changes requested):
The changes look good apart from the missing JSON-RPC API version bump.

And please fix your commit messages to use at least a sentence like `add support to sort items by original title` or something similar.

DaveTBlake commented:
I think that the first commit could also do with a better description too. For example: 
"Add sorting of video items by original title".

DaveTBlake commented on xbmc/filesystem/PluginDirectory.cpp:
```diff
@@ -238,6 +238,12 @@ void CPluginDirectory::AddSortMethod(int handle, SORT_METHOD sortMethod, const s
+	case SORT_METHOD_VIDEO_ORIGINAL_TITLE:
+	case SORT_METHOD_VIDEO_ORIGINAL_TITLE_IGNORE_THE:
+	  {
+        dir->m_listItems->AddSortMethod(SortByOriginalTitle, 556, LABEL_MASKS(labelMask, "%M", labelMask, "%M"), CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_FILELISTS_IGNORETHEWHENSORTING) ? SortAttributeIgnoreArticle : SortAttributeNone);
```
We are working towards making all code to the same standard so new lines of code needs to be to our current style guides rather than copy the style around it. This looks way too long. Run `clang` on the changed/added lines of code.

JimmyS83 replied:
I dont understand. I followed existing format of a file which I changed (SORT_METHOD_VIDEO_SORT_TITLE_IGNORE_THE for reference). This is minor change, not refactor bigger chunks of application like filesystem part. I dont have such intentions..

Montellese replied:
Just wait for the jenkins build to finish and it will automatically leave a comment here with the formatting changes necessary. Then you can apply those to your commits.

Montellese reviewed (approved):
Looks good to me, thanks.

@DaveTBlake do you want to hold this back as it poses the same issue as #19372 concerning JSON-RPC version bumps?

DaveTBlake commented:
> do you want to hold this back as it poses the same issue as #19372 concerning JSON-RPC version bumps?

Yes, would be useful to hold merging this until after 19.2 point release it out.

Montellese reviewed (changes requested):
Looks good in general but I found a few things that need to be addressed. It also looks like you lost the JSON-RPC version bump in your rebase.

Montellese reviewed (changes requested):
Apart from the version bump which needs fixing everything is fine for me now. Thanks for your patience.

phunkyfish commented:
@JimmyS83 you appear to have lost the version bump again.

JimmyS83 commented:
@phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(

phunkyfish commented:
> @phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(

Fair point. I’ve get that pain. Let’s see can we get this merged.

@DaveTBlake you ok with this being merged once the version is added?

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -13313,6 +13313,8 @@ msgctxt "#20375"
 msgid "Unassign content"
 msgstr ""
 
+#: xbmc/video/GUIViewStateVideo.cpp
```
Please add a verbal description to help our translators so they can get their job done.
I mean thise lines starting with `#.` ;-)

JimmyS83 replied:
done

ksooo commented on xbmc/utils/SortUtils.cpp:
```diff
@@ -276,6 +276,22 @@ std::string BySortTitle(SortAttribute attributes, const SortItem &values)
+  if (title.empty())
+    title = values.at(FieldTitle).asString();
+
+  if (attributes & SortAttributeIgnoreArticle)
```
Maybe `if (!title.empty() && (attributes & ...))` ?

JimmyS83 replied:
Are you sure that _SortAttributeIgnoreArticle_ should be dependent on non-empty title? In that case I think _std::string BySortTitle_ must be changed also for consistency

ksooo replied:
what could the result of `SortUtils::RemoveArticles` called with an empty string be? nothing else tan an ampty string, right? So, if title is empty, there should be no need to call `SortUtils::RemoveArticles`. That's the whole idea.

Montellese replied:
I don't think `title` will ever be empty because we have the fallback to use `FieldTitle` and I don't think many people have items without titles in their libraries.

ksooo replied:
Okay, then, Thanks for clarifying, @Montellese

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -13315,6 +13315,7 @@ msgstr ""
 
 #: xbmc/video/GUIViewStateVideo.cpp
 #: xbmc/playlists/SmartPlayList.cpp
+#: There is option to sort list also by Original Title (ListItem.OriginalTitle)
```
nope. the line must start with #. (hashmark, followed by dot)
And I'm honestly not sure this description is actually helpful. You should at least explain what an "original title" is, so a translator gets an idea *what* this string denotes.

Montellese replied:
The term "original title" has been there for a long time and this PR just cleans up the usage comments. I don't see why he has to add a detailed description of what "original title" means.

ksooo replied:
Translators in Weblate may not be familar with our code to look up the details there. "Original title" is quite a generic term - so, a translator might ask: original title of what? Of a movie? A PVR recording? That might make a difference in non-english languages. A comment explaining what this string is actually used for is really very helpful - says somebody who translated many hundreds of strings for Kodi, skins and add-ons.

JimmyS83 replied:
#. Original title of the movie/tvshow 

is fine for you?

I am struggling with better definition of original, as this term is known for me since years (original title vs "country distributor title" which could differ or not )..
![image](https://user-images.githubusercontent.com/7437057/148634724-3459d597-92cd-4a74-9d86-64cf48ee31ae.png)

JimmyS83 commented:
> > @phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(
> 
> Fair point. I’ve get that pain. Let’s see can we get this merged.
> 
> @DaveTBlake you ok with this being merged once the version is added?

I am fine with bump number of JSON. Current is I believe 12.7.0. If somebody really merge this after resolving all other possible things, just tell me wanted version number, I will add it before squash/rebase at the end.

JimmyS83 commented:
I added requested comment for translators, I added again JSON version (which was changed again 2 days ago), but I dont know if anyone merge this before another JSON version bump...

ksooo commented:
@JimmyS83 please fix the clang-format issues, otherwise the PR wont get merged.

JimmyS83 commented:
> @JimmyS83 please fix the clang-format issues, otherwise the PR wont get merged.

I added formatting changes proposed by jenkins.

ksooo commented:
> I added formatting changes proposed by jenkins.

 Almost. ;-) Jenkins is still not happy.

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

JimmyS83 commented:
I think that one machine has problems with disc space or smth, otherwise everything should be OK.

phunkyfish commented:
Jenkins build this please

JimmyS83 commented:
Merge me please :-)

phunkyfish commented:
Thanks for sticking with it @JimmyS83!
